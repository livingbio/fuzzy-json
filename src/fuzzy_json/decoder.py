import json
from functools import wraps
from typing import Any, Callable


def state(fn: Callable[[str, list[str]], str | None]) -> Callable[[str, list[str]], str]:
    @wraps(fn)
    def wrapper(input: str, stack: list[str] = []) -> str:
        try:
            r = fn(input, stack)
            assert r is not None
        except AssertionError:
            raise ValueError("Invalid JSON {input} in {fn.__name__}")
        return r

    return wrapper


def state_start(input: str) -> str:
    return state_root_object(input, ["$"])


@state
def state_root_object(input: str, stack: list[str]) -> str | None:
    input = input.strip()

    if input[0] == "{":
        return input[0] + state_object(input[1:], stack + ["{"])
    return None


@state
def state_finish(input: str, stack: list[str]) -> str | None:
    input = input.strip()

    if not input:
        return ""

    return None


@state
def state_object(input: str, stack: list[str]) -> str | None:
    input = input.strip()

    if input[0] == "}" and stack[-1] == "{":
        return input[0] + state_post_object(input[1:], stack[:-1])

    if input[0] == '"':
        return input[0] + state_property_string(input[1:], stack)
    return None


@state
def state_post_object(input: str, stack: list[str]) -> str | None:
    if stack[-1] in {"{", "["}:
        return state_post_value(input, stack)

    if stack[-1] == "$":
        return state_finish(input, stack)
    return None


@state
def state_post_property(input: str, stack: list[str]) -> str | None:
    input = input.strip()

    if input[0] == ":":
        return input[0] + state_value(input[1:], stack)
    return None


@state
def state_value(input: str, stack: list[str]) -> str | None:
    input = input.strip()

    if input[0] == "f":
        assert input[:5] == "false"
        return input[:5] + state_post_value(input[5:], stack)
    if input[0] == "t":
        assert input[:4] == "true"
        return input[:4] + state_post_value(input[4:], stack)
    if input[0] == "n":
        assert input[:4] == "null"
        return input[:4] + state_post_value(input[4:], stack)
    if input[0].isdigit() or input[0] == "-":
        return input[0] + state_int(input[1:], stack)
    if input[0] == '"':
        return input[0] + state_value_string(input[1:], stack)
    if input[0] == "{":
        return input[0] + state_object(input[1:], stack + ["{"])
    if input[0] == "[":
        return input[0] + state_value(input[1:], stack + ["["])
    if input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_value(input[1:], stack[:-1])

    return None


@state
def state_post_value(input: str, stack: list[str]) -> str | None:
    input = input.strip()

    if input[0] == ",":
        if stack[-1] == "[":
            return input[0] + state_value(input[1:], stack)
        if stack[-1] == "{":
            return input[0] + state_object(input[1:], stack)
        return None
    elif input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_value(input[1:], stack[:-1])
    elif input[0] == "}":
        assert stack[-1] == "{"
        return input[0] + state_post_object(input[1:], stack[:-1])
    return None


@state
def state_value_string(input: str, stack: list[str]) -> str | None:
    for i in range(len(input)):
        if input[i] == '"':
            try:
                return input[: i + 1] + state_post_value(input[i + 1 :], stack)
            except ValueError:
                # NOTE: assume there is missing escape char
                return input[:i] + state_value_string("\\" + input[i:], stack)

        if input[i] == "\\":
            try:
                return input[: i + 1] + state_escape_char(input[i + 1 :], stack + ["v"])
            except ValueError:
                # NOTE: assume there is escaped-unescaped-character
                return input[:i] + state_value_string(input[i + 1 :], stack)

    return None


@state
def state_property_string(input: str, stack: list[str]) -> str | None:
    for i in range(len(input)):
        if input[i] == '"':
            return input[: i + 1] + state_post_property(input[i + 1 :], stack)
        if input[i] == "\\":
            try:
                return input[: i + 1] + state_escape_char(input[i + 1 :], stack + ["p"])
            except ValueError:
                # NOTE: assume there is escaped-unescaped-character
                return input[:i] + state_property_string(input[i + 1 :], stack)

    return None


@state
def state_escape_char(input: str, stack: list[str]) -> str | None:
    if input[0] == "u":
        int(input[1:5], 16)

        if stack[-1] == "v":
            return input[0:5] + state_value_string(input[5:], stack[:-1])
        if stack[-1] == "p":
            return input[0:5] + state_property_string(input[5:], stack[:-1])

        return None

    if input[0] in {"\\", "/", '"', "b", "f", "n", "r", "t"}:
        if stack[-1] == "v":
            return input[0] + state_value_string(input[1:], stack[:-1])
        if stack[-1] == "p":
            return input[0] + state_property_string(input[1:], stack[:-1])
        return None
    return None


@state
def state_int(input: str, stack: list[str]) -> str | None:
    if input[0].isdigit():
        return input[0] + state_int(input[1:], stack)
    if input[0] == ".":
        return input[0] + state_double(input[1:], stack)
    if input[0] == ",":
        if stack[-1] == "[":
            return input[0] + state_value(input[1:], stack)
        if stack[-1] == "{":
            return input[0] + state_object(input[1:], stack)
        return None
    if input[0] == "}":
        assert stack[-1] == "{"
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0].isspace():
        return input[0] + state_post_value(input[1:], stack)
    # NOTE:
    # the original grammar not accept 3e3 (without dot)
    if input[0] in {"e", "E"}:
        return input[0] + state_exponent_sign(input[1:], stack)

    return None


@state
def state_post_int_parent(input: str, stack: list[str]) -> str | None:
    if stack[-1] == "[":
        return state_post_value(input, stack[:-1])
    if stack[-1] == "{":
        return state_post_object(input, stack[:-1])
    return None


@state
def state_double(input: str, stack: list[str]) -> str | None:
    if input[0].isdigit():
        return input[0] + state_double(input[1:], stack)
    if input[0] == ",":
        if stack[-1] == "[":
            return input[0] + state_value(input[1:], stack)
        if stack[-1] == "{":
            return input[0] + state_object(input[1:], stack)
        return None
    if input[0] == "}":
        assert stack[-1] == "{"
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0].isspace():
        return input[0] + state_post_value(input[1:], stack)
    if input[0] in {"e", "E"}:
        return input[0] + state_exponent_sign(input[1:], stack)
    return None


@state
def state_exponent_sign(input: str, stack: list[str]) -> str | None:
    # NOTE:
    # the original grammar not accept 3.0e3 (without sign)
    if input[0].isdigit():
        return input[0] + state_exponent_digits(input[1:], stack)

    if input[0] in {"+", "-"}:
        return input[0] + state_exponent_digits(input[1:], stack)
    return None


@state
def state_exponent_digits(input: str, stack: list[str]) -> str | None:
    if input[0].isdigit():
        return input[0] + state_exponent_digits(input[1:], stack)
    if input[0] == ",":
        if stack[-1] == "[":
            return input[0] + state_value(input[1:], stack)
        if stack[-1] == "{":
            return input[0] + state_object(input[1:], stack)
        return None
    if input[0] == "}":
        assert stack[-1] == "{"
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0].isspace():
        return input[0] + state_post_value(input[1:], stack)
    return None


def repair_json(json_str: str) -> str:
    return state_start(json_str)


def loads(json_str: str, auto_repair: bool = False) -> dict[str, Any]:
    try:
        return json.loads(json_str, strict=False)
    except json.decoder.JSONDecodeError:
        if not auto_repair:
            raise

    try:
        repaired_json = repair_json(json_str)
    except Exception as e:
        raise json.decoder.JSONDecodeError(f"Failed to repair JSON: {e}", json_str, 0)

    return json.loads(repaired_json, strict=False)
