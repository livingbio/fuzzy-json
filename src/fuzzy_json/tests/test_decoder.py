import json
from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion

from ..decoder import loads, repair_json


def test_repaired_json_simple_case(snapshot: SnapshotAssertion) -> None:
    assert snapshot == json.loads(repair_json("{}"))
    assert snapshot == json.loads(repair_json('{"foo": "bar"}'))
    assert snapshot == json.loads(repair_json('{"foo": "bar", "baz": "qux"}'))
    assert snapshot == json.loads(repair_json('{"foo": "bar", "baz": "qux", "quux": "corge"}'))
    # add more nested objects
    assert snapshot == json.loads(repair_json('{"foo": {"bar": "baz"}}'))
    assert snapshot == json.loads(repair_json('{"foo": {"bar": {"baz": "qux"}}}'))
    assert snapshot == json.loads(repair_json('{"foo": {"bar": {"baz": {"qux": "quux"}}}}'))


@pytest.mark.parametrize(
    "test_filename",
    (Path(__file__).parent / "test_data").glob("valid/*.json"),
    ids=lambda x: x.name,
)
def test_repaired_json_vaild_case(snapshot: SnapshotAssertion, test_filename: Path) -> None:
    content = test_filename.read_text()
    parsed_by_std_json = json.loads(content)

    parsed_by_fixed_json = json.loads(repair_json(content))
    assert snapshot == parsed_by_fixed_json
    assert parsed_by_fixed_json == parsed_by_std_json


@pytest.mark.parametrize(
    "test_filename",
    (Path(__file__).parent / "test_data").glob("invalid/*.jsonx"),
    ids=lambda x: x.name,
)
def test_repaired_json_invaild_case(snapshot: SnapshotAssertion, test_filename: Path) -> None:
    content = test_filename.read_text()
    result = loads(content, auto_repair=True)
    assert snapshot == result


def test_repaired_json_invalid_case_special(snapshot: SnapshotAssertion) -> None:
    content = '{"a": "\n"}'
    result = loads(content, auto_repair=True)
    assert snapshot == result
    assert json.loads(content, strict=False) == result


def test_repair_json_fail() -> None:
    with pytest.raises(json.decoder.JSONDecodeError) as _:
        # test that it will raise JSONDecodeError if it can't fix the JSON
        loads("{", auto_repair=True)
