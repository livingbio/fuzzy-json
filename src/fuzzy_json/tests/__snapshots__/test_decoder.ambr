# serializer version: 1
# name: test_repaired_json_invaild_case[case.jsonx]
  dict({
    'body': '''
      Dear valued client,
      
      We are thrilled to announce that XYZ ELECTRONICS LTD. is celebrating its 50th anniversary in the Electrical and Electronics industry. Since our establishment in 1971, we have been committed to delivering exceptional quality, responsibility, and growth.
      
      Throughout the years, XYZ ELECTRONICS has become a leading manufacturer of high-quality magnet wires, offering the most extensive selection in Taiwan. Our range of products has expanded to include electronic components, optical fiber system products, power source switches, "green" products, and wind power solutions.
      
      To cater to the needs of our global clientele, we have established manufacturing plants overseas, ensuring prompt service and the latest techniques.
      
      Under our management strategy of "Energy, Reinnovation, Reliability," we continuously strive to improve and innovate, ensuring energy efficiency and reliability in all our products.
      
      One of our flagship products is the LSFFY Class F Low Temperature Wire. This wire exemplifies our commitment to providing high-quality materials and meeting the stringent requirements of electrical applications.
      
      We invite you to visit our website at https://xyz-electronics.en.taiwantrade.com/ to explore our comprehensive range of products and discover how XYZ ELECTRONICS can elevate your business with its cutting-edge solutions and unwavering commitment to quality.
      
      Thank you for your continued support and partnership. We look forward to celebrating this significant milestone with you.
      
      Best regards,
      [Your Name]
    ''',
    'subject': 'XYZ ELECTRONICS LTD.: Celebrating 50 Years of Excellence in the Electrical and Electronics Industry',
  })
# ---
# name: test_repaired_json_invaild_case[case2.jsonx]
  dict({
    'body': '''
      Dear valued client, 
      
      We are excited to announce the launch of our latest breakthrough product - the Mandelic Acid and Allantoin Acne-Care Calming Ampoule. This innovative serum is set to revolutionize the way we tackle acne, acne scars, and various skin concerns, delivering a transformative skincare experience like never before.
      
      At XXX Beauty Co. Ltd., we are committed to providing high-quality skincare solutions that meet the diverse needs of our customers. The Mandelic Acid and Allantoin Acne-Care Calming Ampoule is the epitome of our dedication to excellence. This exceptional serum is meticulously formulated to target common skin issues and offer a comprehensive approach to achieving clear, radiant, and blemish-free skin.
      
      Harnessing the power of nature's finest ingredients, our ampoule combines mandelic acid and allantoin to ensure optimal results in addressing acne, acne scars, and uneven skin tone. With regular use, this ampoule whitens spots, smooths out acne scars, improves skin absorption rate, and provides a soothing and moisturizing effect. It is specially formulated to calm sensitive skin and relieve redness, making it suitable for all skin types.
      
      The key ingredient, mandelic acid, has been scientifically proven to effectively improve the abnormality of hair follicle keratin, control sebum secretion, and combat metabolic acne. Additionally, allantoin, a powerful soothing agent, further enhances the efficacy of the ampoule, ensuring a truly transformative skincare experience.
      
      We invite you to visit our official website at [insert contact info url] to learn more about the Mandelic Acid and Allantoin Acne-Care Calming Ampoule and its transformative benefits. Discover the power of this exceptional serum and unlock the secret to clear, radiant, and blemish-free skin.
      
      For any media inquiries, please feel free to contact our spokesperson, 陳陳陳, at [insert contact info email]. You can also find more information on our website at [insert contact info url].
      
      Thank you for your continued support and trust in XXX Beauty Co. Ltd. We look forward to providing you with transformative skincare experiences.
      
      Best regards,
      [Your Name]
      XXX Beauty Co. Ltd.
    ''',
    "subject's": 'Introducing the Mandelic Acid and Allantoin Acne-Care Calming Ampoule',
  })
# ---
# name: test_repaired_json_simple_case
  dict({
  })
# ---
# name: test_repaired_json_simple_case.1
  dict({
    'foo': 'bar',
  })
# ---
# name: test_repaired_json_simple_case.2
  dict({
    'baz': 'qux',
    'foo': 'bar',
  })
# ---
# name: test_repaired_json_simple_case.3
  dict({
    'baz': 'qux',
    'foo': 'bar',
    'quux': 'corge',
  })
# ---
# name: test_repaired_json_simple_case.4
  dict({
    'foo': dict({
      'bar': 'baz',
    }),
  })
# ---
# name: test_repaired_json_simple_case.5
  dict({
    'foo': dict({
      'bar': dict({
        'baz': 'qux',
      }),
    }),
  })
# ---
# name: test_repaired_json_simple_case.6
  dict({
    'foo': dict({
      'bar': dict({
        'baz': dict({
          'qux': 'quux',
        }),
      }),
    }),
  })
# ---
# name: test_repaired_json_vaild_case[case0.json]
  dict({
    'age': 30,
    'email': 'johndoe@example.com',
    'name': 'John Doe',
  })
# ---
# name: test_repaired_json_vaild_case[case1.json]
  dict({
    'location': 'New York',
    'person': dict({
      'age': 25,
      'name': 'Alice',
    }),
  })
# ---
# name: test_repaired_json_vaild_case[case10.json]
  dict({
    'message': '''
      This is an escaped string with special characters: "quotes", \backslashes\, and 
      newlines.
    ''',
  })
# ---
# name: test_repaired_json_vaild_case[case11.json]
  dict({
    'unicode_example': 'This is a Unicode string: © OpenAI •',
  })
# ---
# name: test_repaired_json_vaild_case[case12.json]
  dict({
    'body': '''
      Dear valued client,
      
      We are thrilled to announce that XYZ ELECTRONICS LTD. is celebrating its 50th anniversary in the Electrical and Electronics industry. Since our establishment in 1971, we have been committed to delivering exceptional quality, responsibility, and growth.
      
      Throughout the years, XYZ ELECTRONICS has become a leading manufacturer of high-quality magnet wires, offering the most extensive selection in Taiwan. Our range of products has expanded to include electronic components, optical fiber system products, power source switches, "green" products, and wind power solutions.
      
      To cater to the needs of our global clientele, we have established manufacturing plants overseas, ensuring prompt service and the latest techniques.
      
      Under our management strategy of "Energy, Reinnovation, Reliability," we continuously strive to improve and innovate, ensuring energy efficiency and reliability in all our products.
      
      One of our flagship products is the LSFFY Class F Low Temperature Wire. This wire exemplifies our commitment to providing high-quality materials and meeting the stringent requirements of electrical applications.
      
      We invite you to visit our website at https://xyz-electronics.en.taiwantrade.com/ to explore our comprehensive range of products and discover how XYZ ELECTRONICS can elevate your business with its cutting-edge solutions and unwavering commitment to quality.
      
      Thank you for your continued support and partnership. We look forward to celebrating this significant milestone with you.
      
      Best regards,
      [Your Name]
    ''',
    'subject': 'XYZ ELECTRONICS LTD.: Celebrating 50 Years of Excellence in the Electrical and Electronics Industry',
  })
# ---
# name: test_repaired_json_vaild_case[case13.json]
  dict({
    'empty_array': list([
    ]),
    'empty_object': dict({
    }),
    'escape © OpenAI •': False,
    'escape"property': None,
    'int_array': list([
      1,
      2,
      3,
      4,
    ]),
    'int_in_array': list([
      1,
      2,
      3000.0,
    ]),
    'int_in_obj': dict({
      'a': 3,
    }),
    'sci_in_array': list([
      1.2,
      -1.2,
      0.00025,
      dict({
        'a': 0.00025,
        'b': 0.00025,
      }),
    ]),
    'scientific_notation_example': 2500.0,
    'x': list([
      2.5,
    ]),
    'y': dict({
      'a': 2.5,
      'b': 2.5,
    }),
  })
# ---
# name: test_repaired_json_vaild_case[case2.json]
  dict({
    'fruits': list([
      dict({
        'color': 'red',
        'fruit': 'apple',
      }),
      dict({
        'color': 'yellow',
        'fruit': 'banana',
      }),
    ]),
  })
# ---
# name: test_repaired_json_vaild_case[case3.json]
  dict({
    'employees': list([
      dict({
        'id': 101,
        'name': 'Employee 1',
      }),
      dict({
        'id': 102,
        'name': 'Employee 2',
      }),
    ]),
  })
# ---
# name: test_repaired_json_vaild_case[case4.json]
  dict({
    'age': None,
    'email': None,
    'name': 'Jane Smith',
  })
# ---
# name: test_repaired_json_vaild_case[case5.json]
  dict({
    'active': False,
    'status': True,
  })
# ---
# name: test_repaired_json_vaild_case[case6.json]
  dict({
    'float': 3.14,
    'integer': 42,
  })
# ---
# name: test_repaired_json_vaild_case[case7.json]
  dict({
    'colors': list([
      'red',
      'green',
      'blue',
    ]),
  })
# ---
# name: test_repaired_json_vaild_case[case8.json]
  dict({
    'emptyObject': dict({
    }),
  })
# ---
# name: test_repaired_json_vaild_case[case9.json]
  dict({
    'hobbies': list([
      'reading',
      'hiking',
    ]),
    'person': dict({
      'address': dict({
        'city': 'Cityville',
        'street': '123 Main St',
        'zip': '12345',
      }),
      'name': 'Bob',
    }),
  })
# ---
